Add examples with Content-Digest (#1587)
httpwg/http-extensions: draft-ietf-httpbis-digest-headers.md

make use of features of the lower layer in order to provide some integrity
protection; for instance TCP checksums or TLS records [RFC2818].

> This document defines the representation digest integrity mechanism that acts on
> representation data
> and the content digest integrity mechanism that acts instead
> on the conveyed content.

Reviewer: ```suggestion
This document defines two digest integrity mechanisms for HTTP.
First, representation data integrity, which acts on representation data.
Second, content digest integrity, which acts on conveyed content.
```

> They operate independent of transport integrity, offering

Reviewer: ```suggestion
Both mechanisms operate independent of transport integrity, offering
```

# Examples of Unsolicited Digest {#examples-unsolicited}

The following examples demonstrate interactions where a server responds with a
`Digest` or `Content-Digest` fields even though the client did not solicit one using
`Want-Digest` or `Want-Content-Digest`.

Some examples include JSON objects in the content.
For presentation purposes, objects that fit completely within the line-length limits
are presented on a single line using compact notation with no leading space.
Objects that would exceed line-length limits are presented across multiple lines
(one line per key-value pair) with 2 spaced of leading indentation.

Checksum mechanisms defined in this document are media-type agnostic
and do not provide canonicalization algorithms for specific formats.
Examples are calculated inclusive of any space.
While examples can include both fields,

> `Digest` and `Content-Digest` and can be returned independently.

Reviewer: ```suggestion
`Digest` and `Content-Digest` can be returned independently.
```

## Server Returns Full Representation Data {#example-full-representation}

> In this example, the message content conveys a complete representation data,

Reviewer: ```suggestion
In this example, the message content conveys complete representation data,
```

The response `Digest` field-value is calculated over the JSON object
`{"hello": "world"}`, which is not shown because there is no payload
data.

> `Content-Digest` is computed on an empty content.

Reviewer: ```suggestion
`Content-Digest` is computed on empty content.
```

HTTP/1.1 200 OK
Content-Type: application/json
Digest: sha-256=X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=
Content-Digest: sha-256=47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=

~~~

## Server Returns Partial Representation Data

In this example, the client makes a range request and the server
responds with partial content. The `Digest` field-value represents
the entire JSON object `{"hello": "world"}`

> while the `Content-Digest` one is computed on the message content `"hello"`.

Reviewer: ```suggestion
, while the `Content-Digest` field-value is computed on the message content `"hello"`.
```

Accept: application/json
Accept-Encoding: identity
Digest: sha-256=bWopGGNiZtbVgHsG+I4knzfEJpmmmQHf7RHDXA3o1hQ=

> Location: /books/123

Reviewer: Is this intended? If so, the prose above also needs to be changed to remove reference to Location.

Author: @LPardue this was a typo because `Location` is a response header, while this is the request. Is that ok?

Reviewer: Yep, all good

The following examples demonstrate interactions where a client solicits a
`Digest` using `Want-Digest`.

> The behavior of `Content-Digest` and `Want-Content-Digest` is analogous.

Reviewer: ```suggestion
The behavior of `Content-Digest` and `Want-Content-Digest` is identical.
```

`representation metadata` can expose the communication to tampering. For
example, an actor could manipulate the `Content-Type` field-value and cause a
digest validation failure at the recipient, preventing the application from
accessing the representation. Such an attack consumes the resources of both
endpoints. See also {{digest-and-content-location}}.

Digest fields SHOULD always be used over a connection that provides integrity at
the transport layer that protects HTTP fields.

A `Digest` field using NOT RECOMMENDED digest-algorithms SHOULD NOT be used in
signatures.

Using signatures to protect the checksum of an empty representation
allows receiving endpoints to detect if an eventual payload has been stripped or added.

Any mangling of digest fields, including de-duplication of representation-data-digest values
or combining different field values (see Section 5.2 of {{SEMANTICS}})
might affect signature validation.

## Usage in Trailer Fields

Before sending digest fields in a trailer section, the sender
should consider that intermediaries are explicitly allowed to drop any trailer
(see Section 6.5.2 of {{SEMANTICS}}).

> When digest fields are used in a trailer section, the receiver gets there value after the content
> and may thus be tempted to process the data before validating their value.
> It is preferable that data is only be processed after validation.

Reviewer: ```suggestion
When digest fields are used in a trailer section, the field-values are received after the content.
Eager processing of content before the trailer section prevents digest validation, possibly leading to
processing of invalid data.
```

Specification document(s):  {{digest}} of this document

> ## Want-Content-Digest Field Registration

Reviewer: I think this is ok for the time being. I'll follow up with an editorial suggestion to the IANA section once we land these changes to main branch.
